.github/workflows/rust.yml

```diff
@@ -38,62 +87,120 @@ jobs:
-        env:
-            RUSTFLAGS: "-D warnings"
-        run: cargo test --features ${{ matrix.features }} --no-run
+          path: ~/.cargo/registry
+          key: ${{ runner.os }}-cargo-registry-${{ hashFiles('**/Cargo.lock') }}
+      - name: Cache cargo index
+        uses: actions/cache@v1
+        with:
+          path: ~/.cargo/git
+          key: ${{ runner.os }}-cargo-index-${{ hashFiles('**/Cargo.lock') }}
+      - name: Cache cargo build
+        uses: actions/cache@v1
+        with:
+          path: rust/target
+          key: ${{ runner.os }}-cargo-build-target-${{ hashFiles('**/Cargo.lock') }}
 
-      - name: run tests
+      - name: cargo clippy
         working-directory: ./rust
-        env:
-            RUSTFLAGS: "-D warnings"
-        run: cargo test --features ${{ matrix.features }}
+        run: cargo clippy --features ${{ matrix.features }} -- -D warnings
 
-  clippy:
-    name: rust-clippy
+  test:
+    name: rust-test
     runs-on: ubuntu-latest
+
     strategy:
       matrix:
         features: [default]
       fail-fast: true
     steps:
       - uses: actions/checkout@master
+
       - name: install nightly
         uses: actions-rs/toolchain@v1
         with:
-            profile: minimal
-            toolchain: nightly
-            components: clippy
+          profile: minimal
+          toolchain: nightly
+          components: rustfmt, clippy
 
-      - name: cargo clippy
+      - name: Cache cargo registry
+        uses: actions/cache@v1
+        with:
+          path: ~/.cargo/registry
+          key: ${{ runner.os }}-cargo-registry-${{ hashFiles('**/Cargo.lock') }}
+      - name: Cache cargo index
+        uses: actions/cache@v1
+        with:
+          path: ~/.cargo/git
+          key: ${{ runner.os }}-cargo-index-${{ hashFiles('**/Cargo.lock') }}
+      - name: Cache cargo build
+        uses: actions/cache@v1
+        with:
+          path: rust/target
+          key: ${{ runner.os }}-cargo-build-target-${{ hashFiles('**/Cargo.lock') }}
+
+      - name: Build tests
         working-directory: ./rust
-        run: cargo clippy --features ${{ matrix.features }} -- -D warnings
+        env:
+          RUSTFLAGS: "-D warnings"
+        run: cargo test --features ${{ matrix.features }} --no-run
 
-  rustfmt:
-    name: rustfmt
+      - name: Run unit tests
+        working-directory: ./rust
+        env:
+          RUSTFLAGS: "-D warnings"
+        run: cargo test --features ${{ matrix.features }}
+
+  smoke-test:
+    name: rust-smoke-test
     runs-on: ubuntu-latest
+
     strategy:
+      matrix:
+        features: [default]
       fail-fast: true
     steps:
       - uses: actions/checkout@master
 
       - name: install nightly
         uses: actions-rs/toolchain@v1
         with:
-            profile: minimal
-            toolchain: nightly
-            components: rustfmt
+          profile: minimal
+          toolchain: nightly
+          components: rustfmt, clippy
```

Comment:
I don't think we need these two components here do we?

Reply:
fixed